hw/ip/lc_ctrl/rtl/lc_ctrl_state_pkg.sv.tpl

```diff
@@ -10,9 +10,82 @@
 //
 package lc_ctrl_state_pkg;
 <%
-data_width = lc_st_enc.config['secded']['data_width']
-ecc_width  = lc_st_enc.config['secded']['ecc_width']
+  data_width = lc_st_enc.config['secded']['data_width']
+  ecc_width  = lc_st_enc.config['secded']['ecc_width']
+
+  def _to_pascal_case(inp):
+    oup = ''
+    # only convert if input is in snake case format or ALLCAPS
+    if '_' in inp or inp == inp.upper():
+      upper = True
+      for k in inp:
+        if k == '_':
+          upper = True
+        else:
+          oup += k.upper() if upper else k.lower()
+          upper = False
+    # assume its a possibly PascalCased string.
+    # just make sure the first char is capitalized.
+    else:
+      oup = inp[0].upper()
+      if len(inp) > 1:
+        oup += inp[1:]
+
+    return oup
+
+  def _print_state_enum(prefix, name, config):
+    indent = '    '
+    # Determine name length for nice alignment
+    state_len = 0
+    entry_len = 3
+    for state in config[name].keys():
+      state_len = max(state_len, len(_to_pascal_case(state)))
+      for entry in config[name][state]:
+        entry_len = max(entry_len, len(entry))
+    out = ''
+    for i, state in enumerate(config[name].keys()):
+      out += indent + prefix + _to_pascal_case(state)
+      for k in range(state_len - len(_to_pascal_case(state))):
+        out += ' '
```

Comment:
You can do this a bit easier as `out += ' ' * (state_len - len(_to_pascal_case(state)))` (Python does string replication with str * int).

Of course, then you could fold it in with the surrounding lines. I'd consider making out a list of strings and returning `''.join(out)` at the end. This means you can use brackets to make things a bit cleaner to read (it's probably also a bit faster to run, not that this matters here!).

I did that, then a couple of other semi-mechanical tidy-ups:
- define a variable for `_to_pascal_case(state)` to avoid computing it repeatedly
- instead of adding ", " after everything but the last element, add it before everything but the first: equivalent, but easier to check
- build a dictionary like `config[name]`, but keyed by the Pascal-case name, in the first loop to avoid more dictionary lookups.
- iterate over `reversed(cfg[state])` rather than using `[::-1]` (it's equivalent, but a bit clearer)

This is the result. I think it's a bit easier to read. It also made a couple of questionable bits obvious:
- What happens if we have states `Foo_Bar` and `FOO_BAR`? They get normalised to the same Pascal-case name.
- Is `ZRO` really supposed to be spaced differently from other entries? I think probably not (looking at the output), so I removed the special-case there.

```python
  def _print_state_enum(prefix, name, config):
    indent = ' ' * 4

    # Determine name length for nice alignment; build a new dictionary keyed by
    # the Pascal-case version of the state names.
    state_len = 0
    entry_len = 3
    pstates = {}
    for state, state_cfg in config[name].items():
      pstate = _to_pascal_case(state)
      state_len = max(state_len, len(pstate))
      for entry in state_cfg:
        entry_len = max(entry_len, len(entry))

      # Since _to_pascal_case isn't injective, we should check we didn't get
      # any collisions.
      assert pstate not in pstates
      pstates[pstate] = state_cfg

    out = []
    for i, (pstate, state_cfg) in enumerate(pstates.items()):
      if i:
        out.append(',\n')
      out += [indent, prefix, pstate,
              ' ' * (state_len - len(pstate)), ' = {']
      # Need to iterate in reverse since this is a packed value
      for j, entry in enumerate(reversed(state_cfg)):
          if j:
            out.append(', ')

          ename = 'ZRO' if entry == '0' else entry
          out += [' ' * (entry_len - len(ename)), ename]
      out.append('}')

    return ''.join(out)
```

Reply:
Cool, thanks for helping me clean this up.

Regarding the additional cases: its fine if these are collapsed into the same value.
We do not expect any collisions here (so asserting on that is a good sanity check!).

No `ZRO` is not supposed to be spaced differently (it just so happened to not make a difference with `entry_len = 3`).